core/evm_ext.go

```diff
@@ -0,0 +1,107 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import (
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/subnet-evm/constants"
+	"github.com/ava-labs/subnet-evm/core/vm"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	"github.com/ava-labs/subnet-evm/precompile/contracts/deployerallowlist"
+	"github.com/ava-labs/subnet-evm/precompile/modules"
+	"github.com/ava-labs/subnet-evm/precompile/precompileconfig"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	newEVM = vm.NewEVM
+)
+
+func init() {
+	vm.NewEVM = NewEVM
+}
+
+// IsProhibited returns true if [addr] is in the prohibited list of addresses which should
+// not be allowed as an EOA or newly created contract address.
+func IsProhibited(addr common.Address) bool {
+	if addr == constants.BlackholeAddr {
+		return true
+	}
+
+	return modules.ReservedAddress(addr)
+}
+
+type EVM struct {
+	*vm.EVM
+
+	chainConfig *params.ChainConfig
+}
+
+func NewEVM(
+	blockCtx vm.BlockContext, txCtx vm.TxContext, statedb vm.StateDB,
+	chainConfig vm.ChainConfig, config vm.Config,
+) *vm.EVM {
+	customChainConfig, ok := chainConfig.(*params.ChainConfig)
+	if !ok {
+		// If the chainConfig is not a params.ChainConfig, then we can't use the custom
+		// EVM implementation, so we fall back to the default implementation.
+		log.Warn("ChainConfig is not a *params.ChainConfig, falling back to default EVM")
```

Comment:
Why does passing a non-geth type result in using the native geth implementation and vice versa? That seems back to front.

Reply:
Sorry if it was unclear, at the moment I do not consider `github.com/ava-labs/subnet-evm/params` to be a geth implementation (it imports precompiles).